Delete multi-byte related functions from PAL

am11 opened the pull request:
In PAL, delete:
* `IsDBCSLeadByte`; unimplemented.
* `IsDBCSLeadByteEx`; no-op.
* `GetCPInfo`; has hardcoded length of 4, and the only usage checks for 
  `length == 1`.
* `_mbsinc`; has one usage in ilasm, which is simplified for Unix
  (after the removal of above).
* `_mbsninc`; unused.
* `_mbsdec`; usage is inlined as `p--` because `IsDBCSLeadByteEx` is a
  no-op.
* `GetCaseInsensitiveValueA`, handle ANSI cases as Unicode.
* `IsTextUnicode`, single usage in ilasm is inlined.

Fixes #4778.

am11 commented on src/coreclr/src/pal/src/locale/unicode.cpp:
```diff
@@ -260,70 +260,6 @@ GetACP(VOID)
-    ENTRY("IsDBCSLeadByteEx(CodePage=%#x, TestChar=%d)\n", CodePage, TestChar);
-
-    /* Get the lead byte info with respect to the given codepage*/
-    if( !GetCPInfo( CodePage, &cpinfo ) )
```
There are two usages of GetCPInfo left in PAL:
https://github.com/dotnet/runtime/blob/513ade6f40cde36c2663f0fbd6c8b0947439ba28/src/coreclr/src/pal/src/cruntime/mbstring.cpp#L126
https://github.com/dotnet/runtime/blob/513ade6f40cde36c2663f0fbd6c8b0947439ba28/src/coreclr/src/pal/src/cruntime/mbstring.cpp#L183

but logically these conditions will never satisfied as we have hardcoded MaxCharSize=4 in success case.

jkotas replied:
Agree. I would delete `GetCPInfo` and all its uses.

jkotas commented on src/coreclr/src/pal/src/cruntime/mbstring.cpp:
```diff
@@ -70,10 +70,6 @@ _mbsinc(
     else
```
Can these `_mbsinc` methods be deleted too?

am11 commented:
Linker test leg timed out.

am11 commented:
Alpine arm64 `504 Gateway Time-out` response from helix is unrelated.

am11 commented on src/coreclr/src/ilasm/grammar_after.cpp:
```diff
@@ -78,7 +78,11 @@ static Keywords keywords[] = {
+#ifdef HOST_WINDOWS
     return (*pos > 0) ? ++pos : (char *)_mbsinc((const unsigned char *)pos);
+#else
+    return ++pos;
```
Looks like the only usage of `nextcharA` is wrong as it is comparing function pointers instead of result of functions, so this condition always returns true:
https://github.com/dotnet/runtime/blob/342d9d342a440121def32c09735881ba64bd0cdc/src/coreclr/src/ilasm/grammar_after.cpp#L938

jkotas replied:
Yes, this is comparing function pointers; but I do not see why this condition would be always true.

am11 replied:
I was thinking it should do `nextchar(curPos) != nextcharA(curPos)`.

jkotas replied:
The code is using the function pointer as encoding identifier. `nextchar(curPos) != nextcharA(curPos)` would change what this code does.

am11 replied:
Ah ok, I just saw the definition of `nextchar`, it is not a function just a handle declared elsewhere which changes contextually.

jkotas commented on src/coreclr/src/ilasm/grammar_after.cpp:
```diff
@@ -1542,18 +1534,17 @@ char* AsmParse::fillBuff(__in_opt __nullterminated char* pos)
         {
-            if(assem->m_fReportProgress) printf("Source file is ANSI\n\n");
-            penv->pfn_nextchar = nextcharA;
+            if(assem->m_fReportProgress) printf("Source file is ANSI which is not supported. Treating as UTF-8\n\n");
```
I would just print `"Source file is UTF-8\n\n` here (same as in the other path). Pretend that the ANSI encoding support in ilasm never existed.

jkotas commented on src/coreclr/src/ilasm/grammar_after.cpp:
```diff
@@ -931,12 +926,9 @@ int yylex()
                                     {
                                         // if not ANSI, then string is in UTF-8,
```
This comment does not make sense anymore

am11 commented on src/coreclr/src/utilcode/sstring.cpp:
```diff
@@ -224,12 +228,14 @@ int SString::CaseCompareHelperA(const CHAR *buffer1, const CHAR *buffer2, COUNT_
         else
         {
             int length = 1;
+#ifdef HOST_WINDOWS
```
* `CaseCompareHelperA` callers use `GetRawASCII` and it is called for:
  * `REPRESENTATION_{ASCII,ANSI}` from `{Equals,Compare}CaseInsensitive`
  * `REPRESENTATION_ANSI` from `MatchCaseInsensitive`
* `CaseCompareHelper` callers use `GetRawUnicode` and it is called for:
  * `REPRESENTATION_UNICODE` from `{Equals,Compare}CaseInsensitive`
  * `REPRESENTATION_{ASCII,UNICODE}` from `MatchCaseInsensitive`

Should we use `GetRawUnicode` for `CaseCompareHelperA` and delete this conditional block (232-237)?

jkotas replied:
Yes, switch the two `REPRESENTATION_ANSI` cases above to use `GetRawUnicode` and delete this conditional block and `GetCaseInsensitiveValueA`.

jkotas commented on src/coreclr/src/ilasm/grammar_after.cpp:
```diff
@@ -1542,18 +1534,17 @@ char* AsmParse::fillBuff(__in_opt __nullterminated char* pos)
     else
     {
-        if(((penv->curPos[0]&0xFF)==0xEF)&&((penv->curPos[1]&0xFF)==0xBB)&&((penv->curPos[2]&0xFF)==0xBF))
+        if((penv->curPos[0] & 0xFF) == 0xEF && (penv->curPos[1] & 0xFF) == 0xBB && (penv->curPos[2] & 0xFF) == 0xBF)
```
It may be nice to delete `IsItUnicode` and `IsTextUnicode` in the PAL, and just inline the check for UTF16 BOM from `IsTextUnicode` above so that it looks similar to this check for UTF8 BOM. It would fix https://github.com/dotnet/runtime/issues/4778.

am11 replied:
There is a corner case which is not handled by `IsTextUnicode` for UTF-16 LE, i.e. after FF and FE, position 2 and position 3 should not be 0. In terms of inlined version:

```diff
- if(iPutToBuffer >= 2 && (penv->curPos[0] & 0xFF) == 0xFF && (penv->curPos[1] & 0xFF) == 0xFE)
+ if(iPutToBuffer >= 4 && (penv->curPos[0] & 0xFF) == 0xFF && (penv->curPos[1] & 0xFF) != 0xFE && (penv->curPos[2] & 0xFF) != 0x00 && (penv->curPos[3] & 0xFF) != 0x00)
```

because:

```sh
# 0xFFFE followed by 0x00 and 0x00 is UTF-32
$ printf "\xff\xfe\x00\x00" > /tmp/x && file /tmp/x
/tmp/x: Unicode text, UTF-32, little-endian

# any non-zero value after 0xFFFE is UTF-16 LE
$ printf "\xff\xfe\x00\x01" > /tmp/x && file /tmp/x
/tmp/x: Little-endian UTF-16 Unicode text, with no line terminators
```

am11 commented on src/coreclr/src/pal/inc/pal.h:
```diff
@@ -4021,8 +4001,6 @@ PALIMPORT DLLEXPORT int __cdecl _stricmp(const char *, const char *);
 PALIMPORT int __cdecl __iscsym(int);
-PALIMPORT unsigned char * __cdecl _mbsinc(const unsigned char *);
-PALIMPORT unsigned char * __cdecl _mbsninc(const unsigned char *, size_t);
 PALIMPORT unsigned char * __cdecl _mbsdec(const unsigned char *, const unsigned char *);
```
Yup, it is deleted. :)

am11 committed:
Delete multi-byte related functions from PAL

In PAL, delete:
* `IsDBCSLeadByte`; unimplemented.
* `IsDBCSLeadByteEx`; no-op.
* `GetCPInfo`; has hardcoded length of 4, and the only usage checks for
  `length == 1`.
* `_mbsinc`; has one usage in ilasm, which is simplified for Unix
  (after the removal of above).
* `_mbsninc`; unused.
* `_mbsdec`; usage is inlined as `p--` because `IsDBCSLeadByteEx` is a
  no-op.
* `GetCaseInsensitiveValueA`, handle ANSI cases as Unicode.
* `IsTextUnicode`, single usage in ilasm is inlined.

am11 commented:
Deleted`_MBS_SUPPORT` macro (which is always 0 in PAL) and its single remaining usage in https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/coreclr/src/pal/src/safecrt/tsplitpath_s.inl#L105-L114

jkotas commented on src/coreclr/src/ilasm/grammar_after.cpp:
```diff
@@ -1520,19 +1505,15 @@ void AsmParse::ParseFile(ReadStream* stream)
     penv->endPos = penv->curPos + iPutToBuffer;
     if(iPutToBuffer > 128) iPutToBuffer = 128;
-    if(IsItUnicode(penv->curPos,iPutToBuffer,&iOptions))
+    if(iPutToBuffer >= 4 && (penv->curPos[0] & 0xFF) == 0xFF && (penv->curPos[1] & 0xFF) == 0xFE && ((penv->curPos[2] & 0xFF) == 0x00 || (penv->curPos[3] & 0xFF) != 0x00))
```
`(penv->curPos[3] & 0xFF) != 0x00)` case was incorrectly using UTF16 before and it will incorrectly use UTF8 now.  Should this error out instead?

jkotas commented on src/coreclr/src/ilasm/grammar_after.cpp:
```diff
@@ -1520,19 +1505,15 @@ void AsmParse::ParseFile(ReadStream* stream)
 {
     int iPutToBuffer;
-    int iOptions = IS_TEXT_UNICODE_UNICODE_MASK;
     g_uCodePage = CP_ACP;
```
This will be always overwritten by CP_UTF8 below. Simplify?

jkotas commented on src/coreclr/src/pal/src/safecrt/internal_securecrt.h:
```diff
@@ -259,9 +259,6 @@
-#define _MBSDEC(_String, _Current) \
-    _mbsdec((_String), (_Current))
-
 #define _ISMBBLEADPREFIX(_Result, _StringStart, _BytePtr)               \
```
This macro looks unused. Delete it?

jkotas commented on src/coreclr/src/utilcode/makepath.cpp:
```diff
@@ -101,7 +101,7 @@ void MakePath (
                 }
 
 #ifdef _MBCS
```
Is `_MBCS` ever defined? Delete this instead?

am11 replied:
`_MBCS` and `_ismbblead` another unused ~1000 loc, which is removed in ba49728.

jkotas commented on src/coreclr/src/utilcode/sstring.cpp:
```diff
@@ -197,49 +169,21 @@ int SString::CaseCompareHelperA(const CHAR *buffer1, const CHAR *buffer2, COUNT_
         {
-            diff = ch1 - ch2;
-            if  (diff != 0 || stopOnNull)
+            if (!CAN_SIMPLE_UPCASE_ANSI(ch1) || !CAN_SIMPLE_UPCASE_ANSI(ch2))
```
I would delete this condition - it will early for characters <0x20. I think it is better to always take the `SIMPLE_UPCASE_ANSI` path here.

jkotas reviewed (approved):
LGTM otherwise. Thank you!

am11 committed:
Delete _MBCS, _ismbblead and simplifications

am11 committed:
Bring back the ANSI condition

am11 committed:
Delete all _MBCS

jkotas commented:
> Bring back the ANSI condition

When is this needed?

am11 commented:
I had to bring back the condition, as it was failing crossgen during the runtime on all legs. Only reverting that condition fixed the build.

am11 commented:
The error message was bit vague (but error code was consistent across the legs), e.g. from Windows:

_https://dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_apis/build/builds/797301/logs/170_
```
##[error]CROSSGEN-CORELIB(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Build) CrossGen System.Private.CoreLib build failed. Refer to F:\workspace\_work\1\s\src\coreclr\..\..\artifacts\log\CrossgenCoreLib_Windows_NT__x64__Checked.log
  CROSSGEN-CORELIB: "F:\workspace\_work\1\s\src\coreclr\..\..\artifacts\bin\coreclr\Windows_NT.x64.Checked\crossgen.exe" /nologo  /Platform_Assemblies_Paths "F:\workspace\_work\1\s\src\coreclr\..\..\artifacts\bin\coreclr\Windows_NT.x64.Checked\IL" /out "F:\workspace\_work\1\s\src\coreclr\..\..\artifacts\bin\coreclr\Windows_NT.x64.Checked\System.Private.CoreLib.dll" "F:\workspace\_work\1\s\src\coreclr\..\..\artifacts\bin\coreclr\Windows_NT.x64.Checked\IL\System.Private.CoreLib.dll" 
  Error compiling F:\workspace\_work\1\s\src\coreclr\..\..\artifacts\bin\coreclr\Windows_NT.x64.Checked\IL\System.Private.CoreLib.dll:  could not be found. (0x80030002 (STG_E_FILENOTFOUND))
EXEC : error : compilation failed for "F:\workspace\_work\1\s\src\coreclr\..\..\artifacts\bin\coreclr\Windows_NT.x64.Checked\IL\System.Private.CoreLib.dll" (0x80030002) [F:\workspace\_work\1\s\src\coreclr\crossgen-corelib.proj]

```

jkotas commented:
> I had to bring back the condition, as it was failing crossgen during the runtime on all legs. Only reverting that condition fixed the build.

Do you have details about why it was failing? I do not think the method will behave correctly with this condition in all situations.

am11 commented:
The check can be written as:

```diff
         if  (diff != 0 || stopOnNull)
         {
-            if (!CAN_SIMPLE_UPCASE_ANSI(ch1) || !CAN_SIMPLE_UPCASE_ANSI(ch2))
+            if (ch1 == 0 || ch2 == 0)
             {
                 break;
             }
```
crossgen succeeds and CI is happy with both forms. Deleting that inner condition fails crossgen with no helpful exception message and lldb does not catch any backtraces (process just terminates with `0x80030002`).

jkotas commented:
`if (ch1 == 0 || ch2 == 0)` would be better.

am11 committed:
Address CR feedback

am11 committed:
Delete trailing whitespaces from PR file changed

```sh
$ git show HEAD...dotnet/master --name-only --pretty="" |\
    xargs -I{} sh -c 'test -f {} && echo {}' |\
     xargs sed -i '' 's/[[:space:]]*$//'
```

jkotas commented:
Thank you!
